packages/devtools_app/lib/src/debugger/flutter/debugger_controller.dart

```diff
@@ -0,0 +1,314 @@
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/foundation.dart';
+import 'package:vm_service/vm_service.dart';
+
+import '../../globals.dart';
+
+/// Responsible for managing the debug state of the app.
+class DebuggerController {
+  VmService _service;
+
+  StreamSubscription<Event> _debugSubscription;
+
+  IsolateRef isolateRef;
+  List<ScriptRef> scripts;
+
+  final Map<String, Script> _scriptCache = <String, Script>{};
+
+  final _isPaused = ValueNotifier<bool>(false);
+  ValueListenable<bool> get isPaused => _isPaused;
+
+  final _hasFrames = ValueNotifier<bool>(false);
+  ValueNotifier<bool> _supportsStepping;
+  ValueListenable<bool> get supportsStepping {
+    return _supportsStepping ??= () {
+      final notifier = ValueNotifier<bool>(_isPaused.value && _hasFrames.value);
+      void update() {
+        notifier.value = _isPaused.value && _hasFrames.value;
+      }
+
+      _isPaused.addListener(update);
+      _hasFrames.addListener(update);
+      return notifier;
+    }();
+  }
+
+  Event lastEvent;
+
+  final _breakpoints = ValueNotifier<List<Breakpoint>>([]);
+  ValueListenable<List<Breakpoint>> get breakpoints => _breakpoints;
+
+  final _exceptionPauseMode = ValueNotifier<String>(null);
+  ValueListenable<String> get exceptionPauseMode => _exceptionPauseMode;
+
+  InstanceRef _reportedException;
+
+  void setVmService(VmService service) {
+    _service = service;
+    switchToIsolate(serviceManager.isolateManager.selectedIsolate);
+
+    _debugSubscription = _service.onDebugEvent.listen(_handleIsolateEvent);
+    serviceManager.isolateManager.onSelectedIsolateChanged
+        .listen(switchToIsolate);
+  }
+
+  void switchToIsolate(IsolateRef ref) async {
+    isolateRef = ref;
+
+    _isPaused.value = false;
+
+    _clearCaches();
+
+    if (ref == null) {
+      _breakpoints.value = [];
+      return;
+    }
+
+    final dynamic result = await _service.getIsolate(isolateRef.id);
+    if (result is Isolate) {
+      final Isolate isolate = result;
+
+      if (isolate.pauseEvent != null &&
+          isolate.pauseEvent.kind != EventKind.kResume) {
+        lastEvent = isolate.pauseEvent;
+        _reportedException = isolate.pauseEvent.exception;
+        _isPaused.value = true;
+      }
+
+      _breakpoints.value = isolate.breakpoints;
+
+      _exceptionPauseMode.value = isolate.exceptionPauseMode;
+    }
+  }
+
+  Future<Success> pause() => _service.pause(isolateRef.id);
+
+  Future<Success> resume() => _service.resume(isolateRef.id);
+
+  Future<Success> stepOver() {
+    // Handle async suspensions; issue StepOption.kOverAsyncSuspension.
+    final bool useAsyncStepping = lastEvent?.atAsyncSuspension == true;
+    return _service.resume(isolateRef.id,
+        step: useAsyncStepping
+            ? StepOption.kOverAsyncSuspension
+            : StepOption.kOver);
+  }
+
+  Future<Success> stepIn() {
+    return _service.resume(isolateRef.id, step: StepOption.kInto);
```

Comment:
let's create a DebuggerService class that handles all interactions with the vm_service. This is a pattern we follow for other pages in devtools. The controller has an instance of the service (DebuggerService in this case) through which it interacts with the vm_service.